frontend/app/lib/validate.server.ts

```diff
@@ -4,9 +4,59 @@ import { CornerType, PositionType, SlideAnimationType } from './types.js'
 const rangeError = { message: 'Value has to be between 16 and 24' }
 
 export const walletSchema = z.object({
-  walletAddress: z.string().min(1, { message: 'Wallet address is required' })
+  walletAddress: z
+    .string()
+    .min(1, { message: 'Wallet address is required' })
+    .superRefine(async (url, ctx) => {
+      if (url.length === 0) return
+
+      try {
+        const parsed = new URL(toWalletAddressUrl(url))
+        if (parsed.protocol !== 'https:') {
+          ctx.addIssue({
+            code: z.ZodIssueCode.custom,
+            message: 'Wallet address must use HTTPS protocol'
+          })
+          return
+        }
+
+        await isValidWalletAddress(parsed.toString())
+      } catch {
+        ctx.addIssue({
+          code: z.ZodIssueCode.custom,
+          message: 'Invalid wallet address format'
+        })
+      }
+    })
 })
 
+function toWalletAddressUrl(s: string): string {
+  return s.startsWith('$') ? s.replace('$', 'https://') : s
+}
+
+const isValidWalletAddress = async (
+  walletAddressUrl: string
+): Promise<void> => {
```

Comment:
I'd expect a `is*` function to return a boolean.

```suggestion
): Promise<boolean> => {
```